dnf/cli/commands/history.py

```diff
@@ -219,24 +243,140 @@ def str2transaction_id(s):
+                _('Unexpected type of "version" in "{fn}", string expected.'.format(fn=fn))
+            )
+
+        dnf.transaction_sr.check_version(self.replay_data["version"], fn)
+
+    def replay_transaction(self):
+        fn = self.transaction_filename
+
+        with open(fn, "r") as f:
+            try:
+                self.replay_data = json.load(f)
+            except json.decoder.JSONDecodeError as e:
+                raise dnf.exceptions.Error(
+                    _('Error parsing transaction file "{fn}": {err}.'.format(fn=fn, err=str(e)))
+                )
+
+        self._verify_toplevel_json()
+
+        rpms = self.replay_data.get("rpms", {})
+        if not isinstance(rpms, dict):
+            raise dnf.exceptions.Error(
+                _('Unexpected type of "rpms" in "{fn}", object expected.'.format(fn=fn))
+            )
+
+        groups = self.replay_data.get("groups", {})
+        if not isinstance(groups, dict):
+            raise dnf.exceptions.Error(
+                _('Unexpected type of "groups" in "{fn}", object expected.'.format(fn=fn))
+            )
 
-        tids, merged_tids = self._args2transaction_ids()
+        environments = self.replay_data.get("environments", {})
+        if not isinstance(groups, dict):
+            raise dnf.exceptions.Error(
+                _('Unexpected type of "environments" in "{fn}", object expected.'.format(fn=fn))
+            )
 
+        self._na_reason_cache = {}
+
+        for na, pkg_data in rpms.items():
+            subj = hawkey.Subject(na)
+            nas = subj.get_nevra_possibilities(forms=[hawkey.FORM_NA])
+
+            if len(nas) != 1:
+                raise dnf.exceptions.Error(_('Cannot parse name.arch for package "{na}".').format(na=na))
+
+            parsed_na = nas[0]
+            try:
+                action = pkg_data["action"]
+
+                # validate the reasons before running the transaction and cache them
+                try:
+                    self._na_reason_cache[na] = \
+                        libdnf.transaction.StringToTransactionItemReason(pkg_data["reason"])
+                except IndexError as e:
+                    raise dnf.exceptions.Error(
+                        _('Unexpected value of package reason "{reason}" for package "{pkg}" in "{fn}".')
+                            .format(reason=pkg_data["reason"], pkg=na, fn=fn)
+                    )
+
+                if action in ("Install", "Upgrade", "Downgrade", "Reinstall"):
+                    nevra = "%s-%s.%s" % (parsed_na.name, pkg_data["to"], parsed_na.arch)
+                    if action in ("Upgrade", "Downgrade", "Reinstall"):
+                        from_nevra = "%s-%s.%s" % (parsed_na.name, pkg_data["from"], parsed_na.arch)
+                        query = dnf.subject.Subject(from_nevra).get_best_solution(self.base.sack)["query"]
+                        if not query.installed():
+                            raise dnf.exceptions.Error(
```

Comment:
It is not good approach to stop on the first error. It would be better to collect all problems and then report them.

Reply:
@lukash Please could you reveal what is the philosophy behind? By replay do you want to perform the same operation or do you want to move the system into the same state like on the original machine?

Reply:
I want to perform the same operation.
Trying to get the system into the same state would (in general) require something much more complex - or checking the system was in the exact same state at the beginning of the transaction (which we can do, still, by checking the rpmdb checksum, as we discussed, I just started with this middle ground solution, as I consider it the most practically useful one).